core/src/block_strider/archive_handler.rs

```diff
@@ -111,22 +113,32 @@ struct Inner<S> {
 }
 
 struct ArchiveListener {
-    rx: parking_lot::Mutex<watch::Receiver<u32>>,
+    current_id: AtomicU32,
+    listener: Arc<tokio::sync::Mutex<Option<u32>>>,
 }
 
 impl ArchiveListener {
-    fn new(rx: watch::Receiver<u32>) -> Self {
+    fn new(listener: Arc<tokio::sync::Mutex<Option<u32>>>) -> Self {
         Self {
-            rx: parking_lot::Mutex::new(rx),
+            listener,
+            current_id: Default::default(),
         }
     }
 
-    fn next_update(&self) -> Result<Option<u32>> {
-        let mut rx = self.rx.lock();
-        if rx.has_changed()? {
-            Ok(Some(*rx.borrow_and_update()))
-        } else {
-            Ok(None)
+    async fn next_update(&self) -> Option<u32> {
+        let id = self.listener.lock().await;
+
+        if let Some(id) = id.as_ref() {
+            if *id != self.current_id.load(Ordering::Acquire) {
+                self.current_id.store(*id, Ordering::Release);
+                return Some(*id);
+            }
```

Comment:
nit: you may use `compare_exchange(current_id, new_id, Ordering::AcqRel, Ordering::Acquire)`